car_cleaning_service_without_threads.rb

```diff
@@ -0,0 +1,198 @@
+require 'time'
+
+class Car
+  attr_reader :id
+  attr_accessor :required_time
+
+  def initialize(id, required_time = nil)
+    @id = id
+    @required_time = required_time
+  end
+
+  def to_s
+    @id
+  end
+
+  def to_sym
+    @id.to_sym
+  end
+end
+
+class CarValidator
+  ID_FORMAT = /^([A-Z]){2}([0-9]){2,3}([A-Z]){3}$/
+
+  attr_reader :errors
+
+  def initialize(car, time)
+    @car = car
+    @time = time
+    @errors = []
+  end
+
+  def valid?
+    validate_id
+    validate_time
+    @errors.empty?
+  end
+
+  private
+
+  def validate_id
+    @errors.push("Invalid car identification number.") unless @car.to_s.match(ID_FORMAT)
+  end
+
+  def validate_time
+    @errors.push("Required time cannot be before the current time.") unless @car.required_time.nil? || @car.required_time > @time
+  end
+end
+
+
+class CarCleaningService
+  MAX_CARS_AT_ONCE = 2
+  CAR_CLEANING_TIME_IN_HOURS = 2
+  ONE_HOUR_IN_SECONDS = 3600
+  WEEKDAY_OPENING_HOUR = 8
+  WEEKDAY_CLOSING_HOUR = 18
+  SATURDAY_OPENING_HOUR = 9
+  SATURDAY_CLOSING_HOUR = 17
+  SUNDAY_OPENING_HOUR = 11
+  SUNDAY_CLOSING_HOUR = 16
+
+  attr_reader :time, :queue
+
+  def initialize()
+    @queue = []
+    @cars_ready = {}
+    @cars_in_process = {}
+    @time = Time.parse(Time.now.strftime("%Y-%m-%dT08:00:00%z")) # the initial time of the simulation is the current day at 8:00
+  end
+
+  public
+
+  def add_car(car)
+    # if the client did not specify the time they need the car, I consider it to be the end of the day
+    car.required_time = Time.parse(@time.strftime("%Y-%m-%dT#{today_closing_hour}:00:00%z")) if car.required_time.nil?
+    position_in_queue = -1
+    @queue.each { |car_in_queue|
```

Comment:
use `do...end` for multi lines blocks and `{}` when there is a single line block